File: src/test/java/no/nav/samordning/hendelser/metrics/AppMetricsTests.java
package no.nav.samordning.hendelser.metrics;

import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.autoconfigure.web.servlet.AutoConfigureMockMvc;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.test.web.servlet.MockMvc;

import static org.junit.jupiter.api.Assertions.*;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.get;

@SpringBootTest
@AutoConfigureMockMvc
class AppMetricsTests {

    @Autowired
    private MockMvc mockMvc;

    @Test
    void hendelser_lest_metrics_are_incremented() throws Exception {
        var metricName = "samordning_hendelser_lest";
        assertEquals(0.0, getMetricValue(metricName));
        mockMvc.perform(get("/hendelser?tpnr=0000"));
        assertNotNull(getMetricValue(metricName));
    }

    private Double getMetricValue(String metric) throws Exception {
        var response = mockMvc.perform(get("/actuator/prometheus"))
                .andReturn().getResponse().getContentAsString().split("\n");

        for (var line : response) {
            if (line.contains("#"))
                continue;

            if (line.startsWith(metric)) {
                return Double.parseDouble(line.split(" ")[1]);
            }
        }

        return null;
    }
}

File: src/test/kotlin/no/nav/samordning/hendelser/metrics/AppMetricsTests.kt
package no.nav.samordning.hendelser.metrics

import com.ninjasquad.springmockk.SpykBean
import io.mockk.verify
import no.nav.samordning.hendelser.security.support.ROLE_SAMHANDLER
import org.junit.jupiter.api.Test
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.autoconfigure.web.servlet.AutoConfigureMockMvc
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.security.test.context.support.WithMockUser
import org.springframework.test.web.servlet.MockMvc
import org.springframework.test.web.servlet.request.MockMvcRequestBuilders.get
import org.springframework.test.web.servlet.result.MockMvcResultMatchers.status

@SpringBootTest
@AutoConfigureMockMvc
internal class AppMetricsTests {

    @Autowired
    private lateinit var mockMvc: MockMvc

    @SpykBean
    private lateinit var appMetrics: AppMetrics

    @Test
    @WithMockUser(roles = [ROLE_SAMHANDLER])
    fun hendelser_lest_metrics_are_incremented() {
        mockMvc.perform(get("/hendelser?tpnr=1000"))
                .andExpect(status().isOk)

        verify(exactly = 1) { appMetrics.incHendelserLest(eq("1000"), any()) }
    }

}
